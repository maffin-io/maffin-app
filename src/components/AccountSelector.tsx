import React from 'react';
import { useHotkeys } from 'react-hotkeys-hook';
import Select, {
  SingleValue,
  components,
} from 'react-select';
import { BiSearch } from 'react-icons/bi';
import useSWRImmutable from 'swr/immutable';

import { Account } from '@/book/entities';
import { getAccountsWithPath } from '@/book/queries';

export type AccountSelectorProps = {
  id?: string,
  placeholder?: string,
  onChange?: Function,
  showRoot?: boolean,
  isClearable?: boolean,
  ignoreAccounts?: string[],
  defaultValue?: Account,
  className?: string,
};

export default function AccountSelector(
  {
    placeholder,
    onChange,
    ignoreAccounts,
    id = 'accountSelector',
    showRoot = false,
    isClearable = true,
    defaultValue,
    className = '',
  }: AccountSelectorProps,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  useHotkeys('meta+k', () => {
    ref.current?.focus();
  });

  let { data: accounts } = useSWRImmutable(
    '/api/accounts',
    getAccountsWithPath,
  );

  accounts = accounts || [];
  accounts = accounts.filter(account => !(ignoreAccounts || []).includes(account.type));
  if (!showRoot) {
    accounts = accounts.filter(account => account.type !== 'ROOT');
  }

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={accounts}
      placeholder={placeholder || 'Choose account'}
      onChange={(newValue: SingleValue<Account> | null) => {
        if (onChange) {
          onChange(newValue);
        }
      }}
      isClearable={isClearable}
      defaultValue={defaultValue}
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      getOptionLabel={(option) => option.path}
      getOptionValue={(option) => option.path}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: () => '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none',
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiSearch className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
