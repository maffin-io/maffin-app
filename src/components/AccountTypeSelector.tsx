import React from 'react';
import Select, {
  SingleValue,
  components,
} from 'react-select';
import { BiShuffle } from 'react-icons/bi';

import { Account } from '@/book/entities';

export type AccountTypeSelectorProps = {
  id?: string,
  placeholder?: string,
  onChange?: Function,
  disabled?: boolean,
  ignoreTypes?: string[],
  className?: string,
};

export default function AccountTypeSelector(
  {
    placeholder,
    onChange,
    ignoreTypes = [],
    id = 'accountSelector',
    disabled = false,
    className = '',
  }: AccountTypeSelectorProps,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  const types = Account.TYPES.slice(1).filter(
    type => !ignoreTypes.includes(type),
  ).map(type => ({ type }));

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={types}
      isDisabled={disabled}
      placeholder={placeholder || 'Choose account type'}
      onChange={(newValue: SingleValue<{ type: string }> | null) => {
        if (onChange) {
          onChange(newValue?.type);
        }
      }}
      isClearable
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      getOptionLabel={(option) => option.type}
      getOptionValue={(option) => option.type}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: ({ isDisabled }) => {
          if (isDisabled) {
            return '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-700 !border-none';
          }
          return '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none';
        },
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiShuffle className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
