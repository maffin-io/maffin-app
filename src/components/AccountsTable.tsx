import React from 'react';
import { ColumnDef } from '@tanstack/react-table';
import Link from 'next/link';
import { BiCircle, BiSolidRightArrow, BiSolidDownArrow } from 'react-icons/bi';
import classNames from 'classnames';

import Table from '@/components/Table';
import type { AccountsTree } from '@/types/accounts';

export type AccountsTableProps = {
  tree: AccountsTree,
};

export default function AccountsTable(
  {
    tree,
  }: AccountsTableProps,
): JSX.Element {
  return (
    <Table<AccountsTree>
      columns={columns}
      data={tree.children}
      initialSort={{ id: 'total', desc: true }}
      showHeader={false}
      showPagination={false}
      tdClassName="p-2"
      getSubRows={row => row.children}
    />
  );
}

const columns: ColumnDef<AccountsTree>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    enableSorting: false,
    header: '',
    cell: ({ row }) => (
      <div
        className={`flex pl-${row.depth * 4}`}
      >
        {row.getCanExpand() ? (
          <button
            type="button"
            onClick={() => row.toggleExpanded()}
          >
            {(
              row.getIsExpanded()
                ? <BiSolidDownArrow className="mr-1 text-xs opacity-20" />
                : <BiSolidRightArrow className="mr-1 text-xs opacity-20" />
            )}
          </button>
        ) : (
          <button
            type="button"
            className="cursor-default"
          >
            <BiCircle className="mr-1 text-xs opacity-20" />
          </button>
        )}
        <Link
          className={classNames('badge hover:text-slate-300', {
            'bg-green-500/20 text-green-300': row.original.account.type === 'INCOME',
            'bg-red-500/20 text-red-300': row.original.account.type === 'EXPENSE',
            'bg-cyan-500/20 text-cyan-300': ['ASSET', 'BANK'].includes(row.original.account.type),
            'bg-orange-500/20 text-orange-300': row.original.account.type === 'LIABILITY',
            'bg-violet-500/20 text-violet-300': ['STOCK', 'MUTUAL'].includes(row.original.account.type),
          })}
          href={`/dashboard/accounts/${row.original.account.guid}`}
        >
          {row.original.account.name}
        </Link>
      </div>
    ),
  },
  {
    accessorFn: (row: AccountsTree) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];
