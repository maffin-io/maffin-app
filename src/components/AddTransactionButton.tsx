import React from 'react';
import { BiPlusCircle } from 'react-icons/bi';

import TransactionForm from '@/components/forms/transaction/TransactionForm';
import { Account } from '@/book/entities';
import Modal from '@/components/Modal';
import Tooltip from '@/components/Tooltip';

type AddTransactionButtonProps = {
  account: Account,
};

export default function AddTransactionButton(
  {
    account,
  }: AddTransactionButtonProps,
): JSX.Element {
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);
  const disabled = ['INCOME', 'EXPENSE'].includes(account.type);

  return (
    <>
      <Modal
        title={`${account.path} transaction`}
        open={isModalOpen}
        setOpen={setIsModalOpen}
      >
        <TransactionForm
          onSave={() => { setIsModalOpen(false); }}
          account={account}
        />
      </Modal>
      <Tooltip
        text="Add transactions from accounts that are not INCOME/EXPENSE/STOCK/MUTUAL"
        show={disabled}
      >
        <button
          type="button"
          className="btn-primary"
          disabled={disabled}
          onClick={() => setIsModalOpen(!isModalOpen)}
        >
          <BiPlusCircle className="mr-1" />
          Add Transaction
        </button>
      </Tooltip>
    </>
  );
}
