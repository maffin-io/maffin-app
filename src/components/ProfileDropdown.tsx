import Link from 'next/link';
import Image from 'next/image';
import React from 'react';
import { BiLogOut, BiImport } from 'react-icons/bi';

import type BookStorage from '@/apis/BookStorage';
import { useUser, useBookStorage } from '@/hooks';

export default function ProfileDropdown(): JSX.Element {
  const [bookStorage] = useBookStorage();
  const { user } = useUser();

  const fileImportInput = React.useRef<HTMLInputElement>(null);

  return (
    <div className="group relative h-full pl-3 pr-4 rounded-t-sm hover:bg-gunmetal-800">
      <button
        type="button"
        className="flex h-full w-full pl-3 pr-4 items-center"
      >
        {
          user && user.image !== '' && (
            <Image
              src={user.image}
              width={24}
              height={24}
              referrerPolicy="no-referrer"
              className="rounded-full mx-2 w-8 h-8"
              alt="user"
            />
          )
        }
        <div>
          <p className="font-semibold capitalize">
            {user?.name || '...'}
          </p>
          <p className="text-xs">
            {user?.email || ''}
          </p>
        </div>
      </button>
      <ul className="absolute w-40 hidden bg-gunmetal-700 rounded-b-sm py-1 group-hover:block">
        <li className="px-3 py-2 text-sm hover:bg-gunmetal-800">
          <button
            id="menu-item-0"
            type="button"
            role="menuitem"
            tabIndex={-1}
            className="link inline-block w-full whitespace-nowrap"
            onClick={() => fileImportInput.current !== null && fileImportInput.current.click()}
          >
            <BiImport className="inline-block align-middle mr-1" />
            <span className="inline-block align-middle">Import</span>
          </button>
          <input
            type="file"
            ref={fileImportInput}
            onChange={(e) => importBook(e, bookStorage)}
            style={{ display: 'none' }}
          />
        </li>
        <li className="block px-3 py-2 text-sm hover:bg-gunmetal-800">
          <Link
            id="menu-item-1"
            role="menuitem"
            tabIndex={-1}
            href="/user/logout"
            className="inline-block w-full whitespace-nowrap"
          >
            <BiLogOut className="inline-block align-middle mr-1" />
            <span className="inline-block align-middle">Logout</span>
          </Link>
        </li>
      </ul>
    </div>
  );
}

async function importBook(
  event: React.ChangeEvent<HTMLInputElement>,
  bookStorage: BookStorage | null,
) {
  if (event.target.files !== null && event.target.files[0] !== null) {
    const fileReader = new FileReader();
    fileReader.onload = async (loadEvent) => {
      if (loadEvent.target !== null && loadEvent.target.result !== null) {
        if (bookStorage === null) {
          throw new Error('Failed saving, bookStorage is null');
        }
        const rawBook = new Uint8Array(loadEvent.target.result as ArrayBuffer);
        await bookStorage.save(rawBook);
        // TODO: need to re-render the page and reload datasource!
      }
    };

    fileReader.readAsArrayBuffer(event.target.files[0]);
  }
}
