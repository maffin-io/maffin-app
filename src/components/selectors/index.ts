export { default as AccountSelector } from './AccountSelector';
export { default as CommoditySelector } from './CommoditySelector';
export { default as AccountTypeSelector } from './AccountTypeSelector';
export { default as NamespaceSelector } from './NamespaceSelector';

export type { AccountSelectorProps } from './AccountSelector';
export type { CommoditySelectorProps } from './CommoditySelector';
export type { AccountTypeSelectorProps } from './AccountTypeSelector';
