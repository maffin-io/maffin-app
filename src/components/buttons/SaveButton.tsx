import React from 'react';
import { BiCloudUpload, BiLoader, BiWifiOff } from 'react-icons/bi';
import { useQuery } from '@tanstack/react-query';
import classNames from 'classnames';

import { DataSourceContext } from '@/hooks';
import { IS_PAID_PLAN } from '@/helpers/env';
import { useOnline } from '@/hooks/state';
import { UpgradeTooltip } from '@/components/tooltips';

export default function SaveButton(): JSX.Element {
  const { data: isSaving } = useQuery({
    queryKey: ['state', 'isSaving'],
    queryFn: () => false,
    networkMode: 'always',
  });
  const { isLoaded, save } = React.useContext(DataSourceContext);
  const { isOnline } = useOnline();

  if (!isLoaded) {
    return (
      <button
        type="button"
        className="btn btn-primary"
      >
        ...
      </button>
    );
  }

  return (
    <>
      <button
        id="save-button"
        type="button"
        className={classNames(
          'btn btn-primary',
          {
            'btn-danger': !isOnline,
          },
        )}
        disabled={isSaving || !IS_PAID_PLAN || !isOnline}
        data-tooltip-id="upgrade-tooltip"
        onClick={async () => {
          await save();
        }}
      >
        {
          (
            isSaving
            && isOnline
            && (
            <>
              <BiLoader className="mr-1 animate-spin" />
              <span>Saving...</span>
            </>
            )
          ) || (
            !isOnline
            && (
              <>
                <BiWifiOff className="mr-1" />
                <span>Save</span>
              </>
            )
          ) || (
            <>
              <BiCloudUpload className="mr-1" />
              <span>Save</span>
            </>
          )
        }
      </button>
      {
        !IS_PAID_PLAN
        && <UpgradeTooltip />
      }
    </>
  );
}
