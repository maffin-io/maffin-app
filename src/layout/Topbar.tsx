'use client';

import React from 'react';
import { useRouter } from 'next/navigation';

import ProfileDropdown from '@/components/ProfileDropdown';
import SaveButton from '@/components/buttons/SaveButton';
import { AccountSelector } from '@/components/selectors';
import type { Account } from '@/book/entities';

export default function Topbar(): JSX.Element {
  const router = useRouter();

  return (
    <div className="grid grid-cols-12 bg-gunmetal-700 h-20 fixed z-10 top-0 px-6 left-20 right-0">
      <div className="col-span-3">
        <AccountSelector
          id="globalSearch"
          className="py-5 pl-1"
          placeholder="Search (cmd + k)..."
          onChange={(selected: Account | null) => {
            if (selected) {
              router.push(`/dashboard/accounts/${selected.guid}`);
            }
          }}
        />
      </div>

      <div className="col-span-1 col-end-11 h-20 mb-0">
        <div className="flex h-full justify-center items-center">
          <SaveButton />
        </div>
      </div>

      <div className="col-span-2 col-end-13 h-20">
        <ProfileDropdown />
      </div>
    </div>
  );
}
