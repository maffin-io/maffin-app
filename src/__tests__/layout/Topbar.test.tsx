import React from 'react';
import { render } from '@testing-library/react';

import Topbar from '@/layout/Topbar';
import { AccountSelector } from '@/components/selectors';
import SaveButton from '@/components/SaveButton';
import ProfileDropdown from '@/components/ProfileDropdown';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));
// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/navigation'), 'useRouter');

jest.mock('@/components/ProfileDropdown', () => jest.fn(
  () => <div data-testid="ProfileDropdown" />,
));
const ProfileDropdownMock = ProfileDropdown as jest.MockedFunction<typeof ProfileDropdown>;

jest.mock('@/components/SaveButton', () => jest.fn(
  () => <div data-testid="SaveButton" />,
));
const SaveButtonMock = SaveButton as jest.MockedFunction<typeof SaveButton>;

jest.mock('@/components/selectors/AccountSelector', () => jest.fn(
  () => <div data-testid="accountSelector" />,
));
const AccountSelectorMock = AccountSelector as jest.MockedFunction<typeof AccountSelector>;

describe('Topbar', () => {
  let mockRouterPush: jest.Mock;

  beforeEach(() => {
    mockRouterPush = jest.fn();
    useRouter.mockImplementation(() => ({
      push: mockRouterPush,
    }));
  });

  it('renders as expected', () => {
    const { container } = render(<Topbar />);

    expect(ProfileDropdownMock).toHaveBeenLastCalledWith({}, {});
    expect(SaveButtonMock).toHaveBeenLastCalledWith({}, {});
    expect(AccountSelectorMock).toHaveBeenLastCalledWith(
      {
        id: 'globalSearch',
        className: 'py-5 pl-1',
        onChange: expect.any(Function),
        placeholder: 'Search (cmd + k)...',
      },
      {},
    );
    const { onChange } = AccountSelectorMock.mock.calls[0][0];
    if (onChange) {
      onChange({ guid: '123' });
    }
    expect(mockRouterPush).toHaveBeenCalledWith('/dashboard/accounts/123');
    expect(container).toMatchSnapshot();
  });
});
