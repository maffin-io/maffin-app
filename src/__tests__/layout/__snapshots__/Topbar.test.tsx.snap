// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topbar renders as expected 1`] = `
<div>
  <div
    class="grid grid-cols-12 bg-white dark:bg-dark-700 shadow-sm h-20 fixed z-10 top-0 px-6 left-20 right-0"
  >
    <div
      class="col-span-3"
    >
      <div
        data-testid="accountSelector"
      />
    </div>
    <div
      class="col-span-9"
    >
      <div
        class="flex h-full items-center justify-end"
      >
        <div
          class="px-2"
        >
          <div
            data-testid="SaveButton"
          />
        </div>
        <div
          class="px-2"
        >
          <div
            data-testid="ThemeButton"
          />
        </div>
        <div
          data-testid="ProfileDropdown"
        />
      </div>
    </div>
  </div>
</div>
`;
